### PR TITLE
Fix issue in read_bam implementation, it got the wrong contig name

### PR DESCRIPTION
**Pull request by gamazeps**
`pysam`'s [`rname`](https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignedSegment.rname) actually returns the index of the contig instead of the contig name (which is poor naming on their end, but they deprecated that API).

We should use `reference_name` instead to get the contig name

Also added 2 integration tests for reading `bam` and `sam` files, None for `cram` because it would require a `fasta` and I don't have a dummy one to use.

Fixed a bug in the json serialization, where if the tag was a list it ended up not being serializable (probably an edge case, but the dummy samtools test file had one)

Also added filetype inference to be able to read sam and cram files

### CODE REVIEW

File: bioframe/io/fileops.py [hunk 1]
```diff
@@ -231,35 +232,55 @@ def read_pairix(
     return df
 
 
-def read_bam(fp, chrom=None, start=None, end=None):
+def read_alignment(fp, chrom=None, start=None, end=None):
     """
-    Read bam records into a DataFrame.
+    Read alignemnt records into a DataFrame.
```
Comment (nvictus) [inline bioframe/io/fileops.py:237]:
```suggestion
    Read alignment records into a DataFrame.
```

File: bioframe/io/fileops.py [hunk 2]
```diff
@@ -231,35 +232,55 @@ def read_pairix(
...
+                    s.reference_name,
+                    s.pos,
+                    s.mapq,
+                    s.cigarstring if s.mapq != 0 else np.nan,
+                    s.rnext,
+                    s.pnext,
+                    s.tlen,
+                    s.seq,
+                    s.qual,
+                    json.dumps(OrderedDict(tags)),
```
Comment (nvictus) [inline bioframe/io/fileops.py:269]: I think we can remove `OrderedDict` now that Python dicts preserve order natively.

**Review by nvictus — APPROVED** [review]
LGTM. My only suggestion is to remove the `OrderedDict` but it's not urgent.

File: bioframe/io/fileops.py (hunk 2 above)
Comment (gamazeps) [reply bioframe/io/fileops.py]: Done

**Comment by gamazeps** [comment]
All done :)
